src/Chainweb/Pact/SPV.hs

```diff
@@ -311,6 +332,165 @@ extractProof True (Object (ObjectMap o) _ _ _) = case M.lookup "proof" o of
+      putBS domainHash
+      putBS mmimSignedCheckpointRoot
+      Binary.put mmimSignedCheckpointIndex
+      putBS messageId
+  let
+    digest = keccak256 $ BL.toStrict $ Binary.runPut $ do
+      putBS ethereumHeader
+      putBS hash'
+
+  let recoverAddress = recoverHexAddress digest
+  addresses <- mapM recoverAddress mmimSignatures
+
+  let verificationAddresses = take threshold $ catMaybes addresses
+  let verifyStep (_, vals) signer = case V.elemIndex signer vals of
+        Just i -> let newV = snd $ V.splitAt (i + 1) vals in (True, newV)
+        Nothing -> (False, V.empty)
+  let verified = fst $ foldl verifyStep (False, validators) verificationAddresses
+
+  let TokenMessageERC20{..} = hmTokenMessage
+  let
+    encodedSender = encodeHex hmSender
+    encodedRecipient = encodeHex hmRecipient
+    hmObj = obj
+          [ ("version", tLit $ LInteger $ toInteger hmVersion)
+          , ("nonce", tLit $ LInteger $ toInteger hmNonce)
+          , ("originDomain", tLit $ LInteger $ toInteger hmOriginDomain)
+          , ("sender", tStr $ asString encodedSender)
+          , ("destinationDomain", tLit $ LInteger $ toInteger hmDestinationDomain)
+          , ("recipient", tStr $ asString encodedRecipient)
+          , ("tokenMessage", obj
+              ([ ("recipient", tStr $ asString tmRecipient)
+               , ("amount", tLit $ LDecimal $ wordToDecimal tmAmount)
+              ])
+            )
+          ]
+  pure $ mkObject [ ("message", hmObj), ("messageId", tStr $ asString $ encodeHex messageId), ("verified", tLit $ LBool verified) ]
+
+recoverAddressValidatorAnnouncement :: Text -> Text -> ExceptT Text IO (Object Name)
+recoverAddressValidatorAnnouncement storageLocation sig = do
+  signatureBinary <- case decodeHex sig of
+          Right s -> pure s
+          Left e -> throwError $ Text.pack $ "Decoding of signature failed: " ++ e
+  domainHash <- case decodeHex domainHashHex of
+          Right s -> pure s
+          Left e -> throwError $ Text.pack $ "Decoding of domainHashHex failed: " ++ e
+
+  let
+    hash' = BS.fromShort $ _getBytesN $ _getKeccak256Hash $ keccak256 $ BL.toStrict $ Binary.runPut $ do
+      putBS domainHash
+      putBS $ Text.encodeUtf8 storageLocation
+
+  let
+    announcementDigest = keccak256 $ BL.toStrict $ Binary.runPut $ do
+      putBS ethereumHeader
+      putBS hash'
+
+  let recoverAddress = recoverHexAddress announcementDigest
+  address <- recoverAddress signatureBinary
+  let addr = fmap (tStr . asString) $ address
+
+  case addr of
+    Just a -> return $ mkObject [ ("address", a) ]
+    Nothing -> throwError "Failed to recover address"
+
+encodeHyperMessage :: Object Name -> ExceptT Text IO (Object Name)
+encodeHyperMessage o = do
+  let
+    om = _objectMap $ _oObject o
+    tokenMessage = om ^? at "tokenMessage" . _Just . to (\case { (TObject o' _) -> Just o'; _ -> Nothing }) . _Just
```

Comment:
We have all prisms available for `Term` and `Literal` - you should be able to extract what you need without casing: https://github.com/kadena-io/pact/blob/9d8d0975d58fd811404130a57724bfb5a2f2af80/src/Pact/Types/Term.hs#L1085

Same for below.